### ISSUE TITLE
OpenMW Shaders have large speed issue

### ISSUE THREAD

**Issue by ptitSeb** [ptitSeb/gl4es#166]
Also with current gl4es framerate is really bad if shaders are enabled, like 1fps with newest, 15-20 fps with 1.1.2
https://youtu.be/HxFM2vX2kP0

_Originally posted by @Sisah2 in https://github.com/ptitSeb/gl4es/issues/45#issuecomment-619831107_

**Comment by Sisah2** [comment]
So it must be something between jan 19 and feb 17. Do you have idea now?

**Comment by ptitSeb** [comment]
Maybe that: 72d594723778b3f81d10e88457e7f422dd4610d4 (jan 19)
or 5dc053980b01da7f574ea8e8506994bde21e0c6b (easy to test, force `LIBGL_GL=20`)
or 436ed69eb5ee1b9bc0aef6bb35432f409e448c38 (easy to test, force `LIBGL_FLOAT=0`)
or 62c9db85746760d16f17c8a31b0e95384bd15c3d (feb 17)

**Comment by Sisah2** [comment]
Feb 9 is still fast, so it must be something after that.

**Comment by ptitSeb** [comment]
So the FLOAT or the 16 TMUs...

**Comment by Sisah2** [comment]
436ed69... is still fast, so i hope its the last one

**Comment by Sisah2** [comment]
62c9db8 crashing for me, so im not sure if try clean build or some later commit

**Comment by ptitSeb** [comment]
try 1 commit later

**Comment by Sisah2** [comment]
e914cad crashing too.

**Comment by Sisah2** [comment]
so crashes are fixed with dba782 and this build is slow, so i think its 62c9db8 what causing this issue.

**Comment by ptitSeb** [comment]
Ok.
Just a try, with current version of gl4es, in `src/glx/hardext.c` line 322 before the `if`, insert
```
if(hardext.maxtex>8) hardext.maxtex=8;
```
To try see if limiting maxtex to 8 solve the slowiness issue, or if it's something inside gl4es code.

**Comment by Sisah2** [comment]
so i added it like this
    if(hardext.maxtex>8) hardext.maxtex=8;
    if (hardext.esversion==1) {...
and its still clow.
can you help where to set that LIBGL_DBGSHADERCONV=1? i tried it everywhere and don get any output. so i commented out ifs around printf in shaderconv.c but the output look incomplete for me.

**Comment by ptitSeb** [comment]
For LIBGL_DBGSHADERCONV, you can force the thing to : go in `src/gl/init.c` and line 579, replace 
```
globals4es.dbgshaderconv=ReturnEnvVarIntDef("LIBGL_DBGSHADERCONV",0);
```
with
```
globals4es.dbgshaderconv=1;//ReturnEnvVarIntDef("LIBGL_DBGSHADERCONV",0);
```
all the traces should be in logcat.

**Comment by Sisah2** [comment]
Thanks, so there is it
https://www.mediafire.com/file/fradsy4ga5zkuek/shaderconvfiltered/file

**Comment by ptitSeb** [comment]
Ok, good.

There are 2 things that I can see just looking quickly in the log:
`LIBGL   : 0:4: S0032: no default precision defined for variable '_gl4es_AlphaRef'`
and
`LIBGL: FPE Custom Fragment shader compile failed: 0:2: P0003: Warning: Extension 'GL_EXT_draw_buffers' not supported`

I'll work on those 2 messages that start with.

**Comment by Sisah2** [comment]
Sorry i did some mistakes, i wanted coment out shadows shaders and accidentaly commented out lightning, so redownload please. but errors you mentioned are still here.
https://www.mediafire.com/file/null/shaderconvfiltered/file

**Comment by ptitSeb** [comment]
Ok, redownloaded... But I'll still focus on thoses 2 errors for now.

**Comment by ptitSeb** [comment]
With this commit, you should not have the `LIBGL: FPE Custom Fragment shader compile failed: 0:2: P0003: Warning: Extension 'GL_EXT_draw_buffers' not supported` anymore.
Can you check if it's gone for you. If yes, does it improved the speed?

**Comment by Sisah2** [comment]
Messages are still here.
04-29 08:33:10.915 23844 24129 I LIBGL   : 0:4: S0032: no default precision defined for variable '_gl4es_AlphaRef'
04-29 08:33:10.924 23844 24129 I LIBGL   : LIBGL: FPE Custom Fragment shader compile failed: 0:2: P0003: Warning: Extension 'GL_EXT_draw_buffers' not supported

noticed some texture problems, like rain are just white squares, and moon have square around himself and bad color (is blue instead of red i think). So fe655e7 didnt solved everything.

**Comment by ptitSeb** [comment]
When openmw start, gl4es should print to the log everything that is detected. Can you copy/paste that here? There is something I don't understand about the GL_EXT_draw_buffers.

For rain/moon: do you have some setting/command line parameter to use on openmw launch to easily test that on my side?

**Comment by Sisah2** [comment]
i just use --resource arg to point to a shaders with fixed int/floats. those textures are bad with default settings es1 or es2 with or without shaders

here is full logcat. openmw.log really doesnt contain anything usefull
https://www.mediafire.com/file/74fycngc0enbre7/logcat/file

**Comment by ptitSeb** [comment]
The logcat doesn't contain what I need.
Initializing gl4es on the pandora gives that:
```
LIBGL: Initialising gl4es
LIBGL: v1.1.3 built on Apr 28 2020 22:36:25
LIBGL: framebuffer output enabled
LIBGL: Using GLES 2.0 backend
LIBGL: loaded: libGLESv2.so
LIBGL: loaded: libEGL.so
LIBGL: Using GLES 2.0 backend
LIBGL: Hardware Limited NPOT detected and used
LIBGL: Extension GL_EXT_blend_minmax detected and used
LIBGL: FBO are in core, and so used
LIBGL: PointSprite are in core, and so used
LIBGL: CubeMap are in core, and so used
LIBGL: BlendColor is in core, and so used
LIBGL: Blend Substract is in core, and so used
LIBGL: Blend Function and Equation Separation is in core, and so used
LIBGL: Texture Mirrored Repeat is in core, and so used
LIBGL: Extension GL_OES_mapbuffer detected
LIBGL: Extension GL_OES_element_index_uint detected and used
LIBGL: Extension GL_OES_packed_depth_stencil detected and used
LIBGL: Extension GL_OES_depth24 detected and used
LIBGL: Extension GL_OES_rgb8_rgba8 detected and used
LIBGL: Extension GL_EXT_multi_draw_arrays detected
LIBGL: Extension GL_EXT_texture_format_BGRA8888 detected and used
LIBGL: Extension GL_OES_depth_texture detected and used
LIBGL: Extension GL_EXT_texture_rg detected and used
LIBGL: Extension GL_OES_texture_float detected and used
LIBGL: Extension GL_OES_texture_half_float detected and used
LIBGL: Extension GL_OES_fragment_precision_high detected and used
LIBGL: Max vertex attrib: 16
LIBGL: Extension GL_OES_standard_derivatives detected and used
LIBGL: Extension GL_OES_get_program detected and used
LIBGL: Number of supported Program Binary Format: 1
LIBGL: Max texture size: 2048
LIBGL: Max Varying Vector: 8
LIBGL: Texture Units: 8/8 (hardware: 8), Max lights: 8, Max planes: 6
LIBGL: Max Color Attachments: 1 / Draw buffers: 1
LIBGL: Hardware vendor is Imagination Technologies
LIBGL: LiveInfo detected, fps will be shown
LIBGL: Targeting OpenGL 2.1
LIBGL: Forcing NPOT support by disabling MIPMAP support for NPOT textures
LIBGL: Not trying to batch small subsequent glDrawXXXX
LIBGL: try to use VBO
LIBGL: glXMakeCurrent FBO workaround enabled
LIBGL: FBO workaround for using binded texture enabled
LIBGL: Current folder is:/media/sda1/sources/x86/risk_of_rain
LIBGL: Loaded a PSA with 126 Precompiled Programs
```
This is what I need to see on your device.

Thating a new game, I have to fill in a new character, get out of the ship, etc... That's long. I know I can use `--skip-menu` and things like `--start="Sadrith Mora"` to skip all this, and I was wondering if you had some good start point to easily test all that.

**Comment by Sisah2** [comment]
for some reason log start with 

04-29 09:11:32.233 28724 28724 I LIBGL   : LIBGL: 
04-29 09:11:32.234 28724 28724 I LIBGL   : LIBGL: 
04-29 09:11:32.234 28724 28724 I LIBGL   : LIBGL: 
04-29 09:11:32.234 28724 28724 I LIBGL   : LIBGL: 
04-29 09:18:33.069 29799 29799 I LIBGL   : LIBGL: 
04-29 09:18:33.070 29799 29799 I LIBGL   : LIBGL: 
04-29 09:18:33.070 29799 29799 I LIBGL   : LIBGL: 
04-29 09:18:33.070 29799 29799 I LIBGL   : LIBGL: 
04-29 09:19:10.121 29945 29945 I LIBGL   : LIBGL: 
04-29 09:19:10.122 29945 29945 I LIBGL   : LIBGL: 
04-29 09:19:10.123 29945 29945 I LIBGL   : LIBGL: 
04-29 09:19:10.123 29945 29945 I LIBGL   : LIBGL: 
04-29 09:19:44.737 30050 30050 I LIBGL   : LIBGL: 
04-29 09:19:44.737 30050 30050 I LIBGL   : LIBGL: 
04-29 09:19:44.741 30050 30050 I LIBGL   : LIBGL: 
04-29 09:19:44.751 30050 30050 I LIBGL   : LIBGL: 
04-29 09:19:44.751 30050 30050 I LIBGL   : LIBGL: 
04-29 09:19:44.751 30050 30050 I LIBGL   : LIBGL: 

then shader conversion messages apear. logcat was running before starting game
i try to find some nice location for testing this staff

**Comment by ptitSeb** [comment]
> according to glextensions apk my phone doesnt support GL_EXT_draw_buffers only GL_EXT_draw_buffers_indexed

Ah, that's interresting. I guess I take the `GL_EXT_draw_buffers` inside `GL_EXT_draw_buffers_indexed` to falsly detect the extension! I'll fix that.

**Comment by ptitSeb** [comment]
Also, with latest commit, I hhave limited the use of 16bits texture on most hardware. I expect this should fix the last texture color issues you had.

**Comment by Sisah2** [comment]
Performance is good now, it looks like both messages disapeared, but rain is still bad and moon dont have that white squere around him, but colors are still swaped i think. also black textures are still here.

**Comment by ptitSeb** [comment]
Good for the performances / error message.

About the bad color, are the menu item ok or also bad?

**Comment by Sisah2** [comment]
menu, trees and particles are fine. but i still compiling that last commit.

**Comment by Sisah2** [comment]
So some textures, moon and rain is still bad.

**Comment by ptitSeb** [comment]
I pushed a commit, fixed a bug in the S3TC handling. Maybe it fixed moon/rain issue?

**Comment by ptitSeb** [comment]
Also, this ticket (speed of shaders) is solved, so if the color is still not solve then create a new ticket.
For the bad reflexion / ground shaders: there is another ticket

This one can be closed I think.
